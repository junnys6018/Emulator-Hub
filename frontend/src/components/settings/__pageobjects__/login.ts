import { findElementBy } from '@/tests/selenium';
import { By } from 'selenium-webdriver';

export class LoginPage {
    private static addProfileLocator() {
        return By.css('button.add-profile');
    }

    private static saveProfileLocator() {
        return By.xpath("//button[text()='Save']");
    }

    private static profileIconLocator() {
        return By.css('button.profile-icon');
    }

    static clickAddProfile() {
        return findElementBy(LoginPage.addProfileLocator()).click();
    }

    static clickSaveButton() {
        return findElementBy(LoginPage.saveProfileLocator()).click();
    }

    static clickProfileIcon() {
        return findElementBy(LoginPage.profileIconLocator()).click();
    }
}
