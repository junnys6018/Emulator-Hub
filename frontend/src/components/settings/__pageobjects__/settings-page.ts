import { By } from 'selenium-webdriver';
import { findElementBy } from '@/tests/selenium';

export class SettingsPanel {
    private static get saveButtonLocator() {
        return By.xpath("//button[text()='Save']");
    }

    private static get resetAllButtonLocator() {
        return By.xpath("//button[text()='Reset All']");
    }

    static saveButton() {
        return findElementBy(SettingsPanel.saveButtonLocator);
    }

    static resetAllButton() {
        return findElementBy(SettingsPanel.resetAllButtonLocator);
    }
}

export class SettingsPage {
    private static sidePanelButtonLocator(panel: string) {
        return By.xpath(`//div[contains(@class, 'settings__button')]/button[text()='${panel}']`);
    }

    private static get titleLocator() {
        return By.css('h1');
    }

    static getTitle() {
        return findElementBy(SettingsPage.titleLocator).getText();
    }

    static clickGeneralSidePanelButton() {
        return findElementBy(SettingsPage.sidePanelButtonLocator('General')).click();
    }

    static clickNESSidePanelButton() {
        return findElementBy(SettingsPage.sidePanelButtonLocator('NES')).click();
    }

    static clickGBSidePanelButton() {
        return findElementBy(SettingsPage.sidePanelButtonLocator('GB')).click();
    }

    static clickGBCSidePanelButton() {
        return findElementBy(SettingsPage.sidePanelButtonLocator('GBC')).click();
    }

    static clickCHIP8SidePanelButton() {
        return findElementBy(SettingsPage.sidePanelButtonLocator('CHIP 8')).click();
    }
}
