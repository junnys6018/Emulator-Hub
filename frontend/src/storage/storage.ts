import { openDB, DBSchema, IDBPDatabase } from 'idb';
import { generateGuestAccount, UserData } from './user-data';
import { v4 as uuidv4 } from 'uuid';

export interface Record {
    uuid: string;
    age: number;
}

export interface EmulatorHubDB extends DBSchema {
    users: {
        value: UserData;
        key: string;
    };
}

export async function createGuestAccount(db: IDBPDatabase<EmulatorHubDB>) {
    console.log('[INFO] No guest account found, generating one');
    const guestUser = await generateGuestAccount();

    const addGuestAccount = async () => {
        try {
            await db.add('users', guestUser);
        } catch (error) {
            if (error.name === 'ConstraintError') {
                console.warn('[WARN] primary key collision, regenerating uuid');
                guestUser.uuid = uuidv4();
                await addGuestAccount();
            } else {
                throw error;
            }
        }
    };
    await addGuestAccount();

    localStorage.setItem('guest-uuid', guestUser.uuid);
    // Dispatch a storage event, this event is not triggered on the window object that modified localStorage
    // so we do it manually
    const event = new StorageEvent('storage', {
        key: 'guest-uuid',
        newValue: guestUser.uuid,
    });
    window.dispatchEvent(event);
}

export async function initializeDatabase() {
    const db = await openDB<EmulatorHubDB>('emulator-hub', 1, {
        upgrade(db, oldVersion) {
            switch (oldVersion) {
                case 0:
                    // database didnt exist, initialize schema
                    console.log('[INFO] Upgrading database to version 1');
                    db.createObjectStore('users', { keyPath: 'uuid' });
            }
        },
        //blocked() {},
        blocking() {
            db.close();
            alert('The application is outdated, please reload the page');
        },
        terminated() {
            alert('[FATAL] database connection was unexpectedly closed');
        },
    });

    // Check if a guest account exists, if not create one
    const guestUuid = localStorage.getItem('guest-uuid');
    if (!guestUuid) {
        await createGuestAccount(db);
    }

    db.close();
}
