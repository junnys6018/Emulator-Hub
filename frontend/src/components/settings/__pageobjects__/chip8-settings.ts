import { By } from 'selenium-webdriver';
import { findElementsBy } from '@/tests/selenium';
import { SettingsPanel } from './settings-page';

export class CHIP8SettingsPanel extends SettingsPanel {
    private static get buttonsLocator() {
        return By.css('div.settings__chip8-grid > *');
    }

    static async keyboardButtons() {
        return findElementsBy(CHIP8SettingsPanel.buttonsLocator);
    }
}
