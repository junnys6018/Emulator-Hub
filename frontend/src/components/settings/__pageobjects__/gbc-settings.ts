import { By } from 'selenium-webdriver';
import { findElementsBy } from '@/tests/selenium';
import { SettingsPanel } from './settings-page';

export class GBCSettingsPanel extends SettingsPanel {
    private static get buttonsLocator() {
        return By.css('div.grid > *');
    }

    static async keyboardButtons() {
        const elements = await findElementsBy(GBCSettingsPanel.buttonsLocator);
        // Select all the keyboard buttons from the grid
        return elements.filter((_element, index) => {
            return index % 3 === 1 && index !== 1;
        });
    }
}
