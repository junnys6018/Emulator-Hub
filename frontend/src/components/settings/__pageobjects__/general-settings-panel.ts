import { By } from 'selenium-webdriver';
import { findElementBy } from '@/tests/selenium';
import { SettingsPanel } from './settings-page';

export class GeneralSettingsPanel extends SettingsPanel {
    private static get showHiddenGamesSwitchLocator() {
        return By.css("input[type='checkbox']");
    }

    static showHiddenGamesSwitch() {
        return findElementBy(GeneralSettingsPanel.showHiddenGamesSwitchLocator);
    }
}
