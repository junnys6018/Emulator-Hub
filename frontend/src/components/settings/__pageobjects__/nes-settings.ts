import { By } from 'selenium-webdriver';
import { findElementsBy } from '@/tests/selenium';
import { SettingsPanel } from './settings-page';

export class NESSettingsPanel extends SettingsPanel {
    private static get buttonsLocator() {
        return By.css('div.settings__controller-grid > *');
    }

    static async keyboardButtons() {
        const elements = await findElementsBy(NESSettingsPanel.buttonsLocator);
        // Select all the keyboard buttons from the grid
        return elements.filter((_element, index) => {
            return index % 3 === 1 && index !== 1;
        });
    }
}
