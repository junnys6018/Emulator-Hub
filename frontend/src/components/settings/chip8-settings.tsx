import React, { Fragment, useCallback, useEffect, useState } from 'react';
import _ from 'lodash';

import { defaultChip8Controls, useUserProfile } from '@/src/storage/user-data';
import { displayKeyCode } from '@/src/util';
import { FaRedo, FaTimes } from 'react-icons/fa';
import { SettingsTitle } from './common';

// TODO: this function still needs some work
function fontSizeLookup(keyCode: string) {
    if (
        keyCode.startsWith('Key') ||
        keyCode.startsWith('Digit') ||
        /* test for function keys */ /^F\d\d?$/.test(keyCode)
    ) {
        return 'text-3xl';
    }

    if (keyCode.length >= 8) {
        return 'text-xs';
    }
    return 'text-base';
}

export default function CHIP8Settings() {
    // The button the mouse is hovered over, if any
    const [hoveredButton, setHoveredButton] = useState<number | null>(null);
    // The button that is currently being edited, if any
    const [editingButton, setEditingButton] = useState<number | null>(null);

    const clearHoveredButton = () => {
        setHoveredButton(null);
    };

    const clearEditingButton = () => {
        setEditingButton(null);
    };

    const [
        {
            settings: { chip8Controls },
        },
        setUserData,
    ] = useUserProfile();

    const [currentSettings, setCurrentSettings] = useState(_.cloneDeep(chip8Controls));

    const settingsChanged = () => {
        return !_.isEqual(currentSettings, chip8Controls);
    };

    // Reorders the input keys for the grid
    const keyboardMap = [1, 2, 3, 12, 4, 5, 6, 13, 7, 8, 9, 14, 10, 0, 11, 15];

    const resetAll = () => {
        // Create a deep copy here, so that the defaults dont get modified
        setCurrentSettings(_.cloneDeep(defaultChip8Controls));
    };

    const onChange = useCallback(
        (action: number, key: string | null) => {
            if (key === null) {
                key = defaultChip8Controls[action];
            }
            currentSettings[action] = key;
            // Create a deep copy, to trigger a re-render
            setCurrentSettings(_.cloneDeep(currentSettings));
        },
        [currentSettings],
    );

    const onSave = () => {
        if (settingsChanged()) {
            setUserData({ settings: { chip8Controls: currentSettings } });
        }
    };

    useEffect(() => {
        const listener = (e: KeyboardEvent) => {
            if (editingButton !== null) {
                onChange(editingButton, e.code);
                setEditingButton(null);
            }
        };

        window.addEventListener('keydown', listener);

        return () => {
            window.removeEventListener('keydown', listener);
        };
    }, [editingButton, onChange]);

    return (
        <Fragment>
            <SettingsTitle title="CHIP 8 Controls" />
            <div className="settings__chip8-grid">
                {keyboardMap.map(index => (
                    <div
                        key={index}
                        className="bg-gray-600 rounded-xl relative mr-7 mb-7 flex items-center justify-center"
                        onMouseLeave={clearHoveredButton}
                        onMouseEnter={() => setHoveredButton(index)}
                    >
                        <span className="absolute top-1 left-2 font-bold text-sm text-gray-400">
                            {index.toString(16).toUpperCase()}
                        </span>
                        {hoveredButton === index && editingButton === null && (
                            <button
                                className="absolute top-2 right-2 focus-visible:outline-none"
                                onClick={() => onChange(index, null)}
                            >
                                <FaRedo size="12px" className="text-primary-500" />
                            </button>
                        )}
                        {editingButton === index && (
                            <button
                                className="absolute top-2 right-2 focus-visible:outline-none"
                                onClick={clearEditingButton}
                            >
                                <FaTimes size="14px" className="text-red-500" />
                            </button>
                        )}
                        <button
                            className={`focus-visible:outline-none ${
                                editingButton === index || (editingButton === null && hoveredButton === index)
                                    ? 'text-primary-500'
                                    : ''
                            } ${fontSizeLookup(currentSettings[index])}`}
                            onClick={e => {
                                // check e.detail to ensure this button was clicked via mouse button
                                // without this check the user cannot assign space or enter as the control
                                if (editingButton === null && e.detail !== 0) {
                                    setEditingButton(index);
                                }
                            }}
                        >
                            {editingButton === index ? '_' : displayKeyCode(currentSettings[index])}
                        </button>
                    </div>
                ))}
            </div>
            <div className="flex mt-auto mb-24">
                <button
                    className={`btn-primary h-10 w-52 mr-20 ${settingsChanged() ? '' : 'disabled'}`}
                    onClick={onSave}
                >
                    Save
                </button>
                <button className="btn-secondary h-10 w-52" onClick={resetAll}>
                    Reset All
                </button>
            </div>
        </Fragment>
    );
}
