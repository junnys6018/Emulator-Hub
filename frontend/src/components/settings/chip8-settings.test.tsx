import _ from 'lodash';
import { driver, resetDriver, elementHasClass, getIndexedDBSettings } from '@/tests/selenium';
import { SettingsPage } from './__pageobjects__/settings-page';
import { CHIP8SettingsPanel } from './__pageobjects__/chip8-settings';
import { By } from 'selenium-webdriver';
import { defaultSettings } from '@/src/storage/user-data';

import React from 'react';
import renderer from 'react-test-renderer';
import CHIP8Settings from './chip8-settings';

beforeEach(async () => {
    await driver.navigate().to('http://localhost:8000/settings');
});

afterEach(async () => {
    resetDriver();
});

jest.setTimeout(15000);

jest.mock('@/src/storage/user-data', () => {
    const originalModule = jest.requireActual('@/src/storage/user-data');
    const useActiveUserProfile = jest.fn();
    useActiveUserProfile.mockReturnValue([
        {
            profileImage: '/image.png',
            userName: 'Guest',
            settings: {
                chip8Controls: [
                    'KeyX',
                    'Digit1',
                    'Digit2',
                    'Digit3',
                    'KeyQ',
                    'KeyW',
                    'KeyE',
                    'KeyA',
                    'KeyS',
                    'KeyD',
                    'KeyZ',
                    'KeyC',
                    'Digit4',
                    'KeyR',
                    'KeyF',
                    'KeyV',
                ],
            },
        },
        jest.fn(),
    ]);
    return {
        __esModule: true,
        ...originalModule,
        useActiveUserProfile,
    };
});

jest.mock('../util/message');
jest.mock('../util/alert');

describe('CHIP8 Settings', () => {
    test('Click on side panel', async () => {
        await SettingsPage.clickCHIP8SidePanelButton();
        expect(await SettingsPage.getTitle()).toBe('CHIP 8 Controls');
    });

    test('Initial state', async () => {
        await SettingsPage.clickCHIP8SidePanelButton();

        const keyboardButtons = await CHIP8SettingsPanel.keyboardButtons();
        expect(keyboardButtons.length).toBe(16);
        const names = ['1', '2', '3', '4', 'Q', 'W', 'E', 'R', 'A', 'S', 'D', 'F', 'Z', 'X', 'C', 'V'];
        for (let i = 0; i < 16; i++) {
            const button = keyboardButtons[i];
            const innerButton = await button.findElement(By.css('button'));
            expect(await innerButton.getText()).toBe(names[i]);
        }

        const saveButton = await CHIP8SettingsPanel.saveButton();
        expect(await elementHasClass(saveButton, 'disabled')).toBe(true);
    });

    test('Changes should be saved', async () => {
        await SettingsPage.clickCHIP8SidePanelButton();

        let keyboardButtons = await CHIP8SettingsPanel.keyboardButtons();
        let innerButton = await keyboardButtons[0].findElement(By.css('button'));
        await innerButton.click();

        const body = driver.findElement(By.css('body'));
        await body.sendKeys('=');

        // fetch buttons again
        keyboardButtons = await CHIP8SettingsPanel.keyboardButtons();
        innerButton = await keyboardButtons[0].findElement(By.css('button'));
        expect(await innerButton.getText()).toBe('EQUAL');

        let saveButton = await CHIP8SettingsPanel.saveButton();
        expect(await elementHasClass(saveButton, 'disabled')).toBe(false);

        // Save settings
        await saveButton.click();

        // check indexeddb
        let result = await getIndexedDBSettings();

        const expected = _.cloneDeep(defaultSettings);
        expected.chip8Controls[1] = 'Equal';

        expect(result).toEqual(expected);

        // Refresh the page
        await driver.navigate().refresh();
        await SettingsPage.clickCHIP8SidePanelButton();

        keyboardButtons = await CHIP8SettingsPanel.keyboardButtons();
        innerButton = await keyboardButtons[0].findElement(By.css('button'));
        expect(await innerButton.getText()).toBe('EQUAL');

        saveButton = await CHIP8SettingsPanel.saveButton();
        expect(await elementHasClass(saveButton, 'disabled')).toBe(true);

        // Reset to defaults
        const resetButton = await CHIP8SettingsPanel.resetAllButton();
        await resetButton.click();
        await saveButton.click();

        // check indexeddb
        result = await getIndexedDBSettings();

        expect(result).toEqual(_.cloneDeep(defaultSettings));
    });

    test('Snapshot', () => {
        const tree = renderer.create(<CHIP8Settings />).toJSON();
        expect(tree).toMatchSnapshot();
    });
});
