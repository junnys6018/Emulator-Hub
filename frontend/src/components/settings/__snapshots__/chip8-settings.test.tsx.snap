// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CHIP8 Settings Snapshot 1`] = `
Array [
  <h2
    className="hidden lg:block text-2xl text-gray-500 mb-7"
  >
    Settings
  </h2>,
  <h1
    className="font-medium text-xl mb-4 lg:text-3xl lg:mb-10"
  >
    CHIP 8 Controls
  </h1>,
  <div
    className="settings__chip8-grid"
  >
    <div
      className="bg-gray-600 rounded-xl relative mr-7 mb-7 flex items-center justify-center"
      onMouseEnter={[Function]}
      onMouseLeave={[Function]}
    >
      <span
        className="absolute top-1 left-2 font-bold text-sm text-gray-400"
      >
        1
      </span>
      <button
        className="focus-visible:outline-none  text-3xl"
        onClick={[Function]}
      >
        1
      </button>
    </div>
    <div
      className="bg-gray-600 rounded-xl relative mr-7 mb-7 flex items-center justify-center"
      onMouseEnter={[Function]}
      onMouseLeave={[Function]}
    >
      <span
        className="absolute top-1 left-2 font-bold text-sm text-gray-400"
      >
        2
      </span>
      <button
        className="focus-visible:outline-none  text-3xl"
        onClick={[Function]}
      >
        2
      </button>
    </div>
    <div
      className="bg-gray-600 rounded-xl relative mr-7 mb-7 flex items-center justify-center"
      onMouseEnter={[Function]}
      onMouseLeave={[Function]}
    >
      <span
        className="absolute top-1 left-2 font-bold text-sm text-gray-400"
      >
        3
      </span>
      <button
        className="focus-visible:outline-none  text-3xl"
        onClick={[Function]}
      >
        3
      </button>
    </div>
    <div
      className="bg-gray-600 rounded-xl relative mr-7 mb-7 flex items-center justify-center"
      onMouseEnter={[Function]}
      onMouseLeave={[Function]}
    >
      <span
        className="absolute top-1 left-2 font-bold text-sm text-gray-400"
      >
        C
      </span>
      <button
        className="focus-visible:outline-none  text-3xl"
        onClick={[Function]}
      >
        4
      </button>
    </div>
    <div
      className="bg-gray-600 rounded-xl relative mr-7 mb-7 flex items-center justify-center"
      onMouseEnter={[Function]}
      onMouseLeave={[Function]}
    >
      <span
        className="absolute top-1 left-2 font-bold text-sm text-gray-400"
      >
        4
      </span>
      <button
        className="focus-visible:outline-none  text-3xl"
        onClick={[Function]}
      >
        Q
      </button>
    </div>
    <div
      className="bg-gray-600 rounded-xl relative mr-7 mb-7 flex items-center justify-center"
      onMouseEnter={[Function]}
      onMouseLeave={[Function]}
    >
      <span
        className="absolute top-1 left-2 font-bold text-sm text-gray-400"
      >
        5
      </span>
      <button
        className="focus-visible:outline-none  text-3xl"
        onClick={[Function]}
      >
        W
      </button>
    </div>
    <div
      className="bg-gray-600 rounded-xl relative mr-7 mb-7 flex items-center justify-center"
      onMouseEnter={[Function]}
      onMouseLeave={[Function]}
    >
      <span
        className="absolute top-1 left-2 font-bold text-sm text-gray-400"
      >
        6
      </span>
      <button
        className="focus-visible:outline-none  text-3xl"
        onClick={[Function]}
      >
        E
      </button>
    </div>
    <div
      className="bg-gray-600 rounded-xl relative mr-7 mb-7 flex items-center justify-center"
      onMouseEnter={[Function]}
      onMouseLeave={[Function]}
    >
      <span
        className="absolute top-1 left-2 font-bold text-sm text-gray-400"
      >
        D
      </span>
      <button
        className="focus-visible:outline-none  text-3xl"
        onClick={[Function]}
      >
        R
      </button>
    </div>
    <div
      className="bg-gray-600 rounded-xl relative mr-7 mb-7 flex items-center justify-center"
      onMouseEnter={[Function]}
      onMouseLeave={[Function]}
    >
      <span
        className="absolute top-1 left-2 font-bold text-sm text-gray-400"
      >
        7
      </span>
      <button
        className="focus-visible:outline-none  text-3xl"
        onClick={[Function]}
      >
        A
      </button>
    </div>
    <div
      className="bg-gray-600 rounded-xl relative mr-7 mb-7 flex items-center justify-center"
      onMouseEnter={[Function]}
      onMouseLeave={[Function]}
    >
      <span
        className="absolute top-1 left-2 font-bold text-sm text-gray-400"
      >
        8
      </span>
      <button
        className="focus-visible:outline-none  text-3xl"
        onClick={[Function]}
      >
        S
      </button>
    </div>
    <div
      className="bg-gray-600 rounded-xl relative mr-7 mb-7 flex items-center justify-center"
      onMouseEnter={[Function]}
      onMouseLeave={[Function]}
    >
      <span
        className="absolute top-1 left-2 font-bold text-sm text-gray-400"
      >
        9
      </span>
      <button
        className="focus-visible:outline-none  text-3xl"
        onClick={[Function]}
      >
        D
      </button>
    </div>
    <div
      className="bg-gray-600 rounded-xl relative mr-7 mb-7 flex items-center justify-center"
      onMouseEnter={[Function]}
      onMouseLeave={[Function]}
    >
      <span
        className="absolute top-1 left-2 font-bold text-sm text-gray-400"
      >
        E
      </span>
      <button
        className="focus-visible:outline-none  text-3xl"
        onClick={[Function]}
      >
        F
      </button>
    </div>
    <div
      className="bg-gray-600 rounded-xl relative mr-7 mb-7 flex items-center justify-center"
      onMouseEnter={[Function]}
      onMouseLeave={[Function]}
    >
      <span
        className="absolute top-1 left-2 font-bold text-sm text-gray-400"
      >
        A
      </span>
      <button
        className="focus-visible:outline-none  text-3xl"
        onClick={[Function]}
      >
        Z
      </button>
    </div>
    <div
      className="bg-gray-600 rounded-xl relative mr-7 mb-7 flex items-center justify-center"
      onMouseEnter={[Function]}
      onMouseLeave={[Function]}
    >
      <span
        className="absolute top-1 left-2 font-bold text-sm text-gray-400"
      >
        0
      </span>
      <button
        className="focus-visible:outline-none  text-3xl"
        onClick={[Function]}
      >
        X
      </button>
    </div>
    <div
      className="bg-gray-600 rounded-xl relative mr-7 mb-7 flex items-center justify-center"
      onMouseEnter={[Function]}
      onMouseLeave={[Function]}
    >
      <span
        className="absolute top-1 left-2 font-bold text-sm text-gray-400"
      >
        B
      </span>
      <button
        className="focus-visible:outline-none  text-3xl"
        onClick={[Function]}
      >
        C
      </button>
    </div>
    <div
      className="bg-gray-600 rounded-xl relative mr-7 mb-7 flex items-center justify-center"
      onMouseEnter={[Function]}
      onMouseLeave={[Function]}
    >
      <span
        className="absolute top-1 left-2 font-bold text-sm text-gray-400"
      >
        F
      </span>
      <button
        className="focus-visible:outline-none  text-3xl"
        onClick={[Function]}
      >
        V
      </button>
    </div>
  </div>,
  <div
    className="flex mt-auto mb-12 lg:mb-24"
  >
    <button
      className="btn-primary h-10 w-40 lg:w-52 mr-auto lg:mr-20 disabled"
      onClick={[Function]}
    >
      Save
    </button>
    <button
      className="btn-secondary h-10 w-40 lg:w-52 mr-20"
      onClick={[Function]}
    >
      Reset All
    </button>
    <button
      className="btn-secondary h-10 w-40 lg:w-52"
      onClick={[Function]}
    >
      Revert
    </button>
  </div>,
]
`;
